Device-Bound Session Credentials Analysis (#1094)
w3ctag/design-reviews: reviews/dbsc-analysis.md

## Usage

In the proposed design, the browser is given three things during enrollment:

* a “session” resource that it can use for protocol interactions,  
* which resources use cookies that are provided using the protocol, plus  
* the set of cookies that can be produced.

The site is given the public key from the site-specific key pair that the browser holds.

In the proposed design, the browser understands that when it makes a request to one of the resources that participates in the protocol, it is expected to hold refreshed versions of the identified cookies.

These cookies are expected to have very short validity periods. The browser is able to refresh those cookies automatically by interacting with the session resource. The main part of the protocol is the interactions between the browser and that session resource.

Interactions with the session resource is a two-step process. The first is a simple request that requests a fresh challenge, the second posts a signature from the secret key over that challenge, thereby proving to the server that the browser still has access to the key pair.  This response also refreshes any of the affected cookies.

> This adds two round trips of latency every time that a cookie refresh is needed. While some amount of delay is likely unavoidable, having two additional requests is fairly heavyweight.

Author: I don't think that it is unnecessary, though it might be depending on the strength of assurance that a site is seeking.  If you consider access to the key (or TPM) to be in the threat model, then interactivity is a useful property.  The problem there is that the TPM is unaware of the time, so you can't bind the signatures it produces to the clock.  So access to the TPM allows an attacker the option to generate arbitrary "future" signatures, allowing it to time travel as necessary.  The way that is addressed in the proposed design is with a liveness check.  I'll amend the below and add some detail.

> We have an alternative below that doesn't require an interactive exchange (though it could be made interactive).  It also includes a redundant new field in requests that lists a session identifier. That new field could be replaced either with a per-account resource URL parameter or a non-DBSC cookie.

Author: ```suggestion
We have an alternative below that doesn't require an interactive exchange. However, given that TPMs generally don't have a clock, you can't use the clock to ensure freshness.  A non-interactive exchange might have been pre-generated by an attacker who temporarily had access to the TPM, unless it contains fresh entropy from the server.  That's something we address in more detail in the alternative design below, noting that the alternative offers servers more options to combine requests to reduce latency, where the proposal cannot.

The proposal includes a redundant new session identifier field in requests. That new field could be replaced either with a per-account resource URL parameter or a non-DBSC cookie.
```

> This process addresses an important concern about the frequency with which the browser needs to access the secret key.  Sites are able to control how often something is signed by setting the expiration date of the cookies they produce. The browser only needs to use the secret key when cookies expire.  Expirations are naturally limited because servers are unable to set extremely short expiration times without risking the cookies being completely useless to clients with bad clock skew; the granularity of expiration dates for cookies is also extremely limited in expressiveness.

Author: Sorry about the wrapping thing.  It's what gdocs does by default.  Probably due to GFM and its soft-breaking thing.

To complete enrollment, the browser needs to generate a key pair, sign the challenge, and post the signature to the session resource. The content of that response is a JSON document that describes the rest of the protocol: which resources and cookies are governed by this resource.

It’s not clear what it means to have multiple sessions concurrently active. There’s some mention of this, but it’s not fully developed. At least theoretically, there is no need to limit the number of sessions, other than to limit the work that browser and server both do and to reduce the state they both need to track.

## Too Complex

Overall, this creates a new set of interaction paradigms between the browser and websites.  We think that there are easier ways to achieve the same basic goals without too much disruption to the existing cookie handling arrangements. That design is sketched below.

# An Alternative Design

This is a sketch of an alternative approach that is far closer to how the web platform currently handles cookies.

> The core requirement for the usage part is that the site is able to regularly request that the browser demonstrate that it has access to the private key that was registered through the enrollment process.  Ideally, that access is not required for most interactions, because generating and validating a digital signature is somewhat expensive for both client and server.

Reviewer: This contradicts footnote [^2], which claims the client-side cost is trivial. We should be internally consistent.

Another challenge is in demonstrating liveness for the signature. TPMs don’t generally have clocks, so if a device is compromised so that an attacker gains access to the TPM, the attacker could generate an arbitrary number of signatures for future use. However, this requires that the attacker predict the times and URLs where those signatures would be needed. This suggests a similar pattern to solve that potential problem also: the server redirects to a new endpoint with fresh randomness in the URL for signing.

For example, both requirements could be addressed as shown below.  This example is expanded to include the maximum number of exchanges possible to fully illustrate all of the capabilities.

```http
GET /some/resource
Cookie: login=expired
```

Which results in a redirection to a login endpoint, as would be part of a normal centralized login flow (i.e., this would be a perfectly normal part of refreshing cookies):

```http
303 Over Yonder

> Location: /login

Reviewer: This should include the `?prev=/some/resource` query parameter that's needed to redirect back to the source page.

Author: The server knows that.  It's just that servers often ask clients to hold stuff while they are busy.  I thought it would be distracting, but now I see that the opposite is true.  I'll fix it up.

```

Consequently, the client follows the redirect.

```http
GET /login
Cookie: login=expired
```

This resource then makes a call about the freshness of the login cookies and determines that a signature is needed.  It initiates another redirect, including fresh entropy in the URL to guarantee that the client has live TPM access.

```http

> 303 Over There Now

Reviewer: The custom reason-phrases are distracting me, and making me wonder if they have a semantic meaning. They [don't](https://httpwg.org/specs/rfc9112.html#status.line), but we should avoid distractions for other readers.

Location: /login/1EU9jsh07pci6Cgk9Bh0
```

After a request that includes the signed cookie,

```http
GET /login/1EU9jsh07pci6Cgk9Bh0
Cookie: login=expired; signed=ok
Signature-Input: (...)
Signature: :...:
```

That resource then can validate the signature and produce an updated cookie.  And likely redirect back to the original resource.

```http
303 Finally
Location: /some/resource
Set-Cookie: login=refreshed; Secure; HttpOnly; etc=etc
```

Note that the server does not need to refresh the signed cookie.  That cookie could be a stub that only exists to elicit a signature, so it could have a very long lifetime.

> These multi-step arrangements would result in similar amounts of delay as the process in the proposal. This approach is still better, because it follows fairly ordinary cookie handling for the most part. Any additional steps would be discretionary on the part of servers, which could sometimes choose to accept either the extra requests with signatures[^2] or the heightened risk of TPM compromise. Alternatively, servers could streamline the overall process by combining steps, at the cost of additional coordination between the different resources. In comparison, the proposed design makes an extra step unavoidable, so making this discretionary is strictly better.

Reviewer: If our alternative doesn't improve the number of requests, we shouldn't flag the 2 round trips as the biggest problem with the current design. We should instead say something like "we think there's a way to re-use existing cookie patterns" or, "we think there's a way to allow the server to save some round trips if it doesn't need all the security provided by the original proposal".

In fact, I think this alternative produces 3 extra round trips (/some/resource, /login, /login/1EU9jsh07pci6Cgk9Bh0, /some/resource), while the original only added 2 as long as the client has a good-enough clock to anticipate that its cookie will be expired (saving the original guaranteed-to-fail request to /some/resource).

Author: Maybe the focus on delays is wrong, I'll look at it closer.

## Communicating Keys

Enrollment can almost be a side effect of creating and first use of a `Signed` cookie.  The only requirement here is that the browser learns what types of keys are acceptable to the server and that the server learns the public key that the client uses.

Any `Set-Cookie` header that establishes a `Signed` cookie could list the key types in the `Signed` attribute, but the `Accept-Signature` field exists for negotiating the use of signature keys. The server should therefore use `Accept-Signature`.

> The `Cookie` header that the browser subsequently sends will be signed.  That same message can include the public key from the key pair.  That’s usually not something that can be included in the signature as defined in the current RFC. For that, we might define a new `Signature-Public-Key` field to carry the necessary information. We could define a new `Signature` field parameter, but that could be confused with `keyid`.

Reviewer: I'm worried about us saying that "here's an alternative design, but you're still going to need to figure out how to modify the message-signature RFC". If that's not straightforward, they're just going to say their current design is better.

Author: That's not at all what this is saying, it's saying that there is a need to define a way to communicate a public key and that is something the message signature RFC doesn't solve.

That design is sketched below.

# An Alternative Design

This is a sketch of an alternative approach that is closer to how the web platform currently handles cookies.

The core requirement for the usage part is that the site is able to
regularly request that the browser demonstrate that it has access to the private key
that was registered through the enrollment process.
Ideally, that access is not required for most interactions,
because generating and validating a digital signature is somewhat expensive for both client and server.

Justin Richer hinted at an alternative pattern that is worth exploring
in [issue 112](https://github.com/w3c/webappsec-dbsc/issues/112).
[HTTP Message Signatures](https://datatracker.ietf.org/doc/html/rfc9421) is not a directly usable standard,
but more of a framework for applying signatures to content.
A signature that covered cookies, URL, date, and other information might be the core of a more complete design.

## Signed Cookies

That design might include a new `Signed` parameter for cookies in `Set-Cookie`.
That attribute would request that, whenever the cookie is sent to the server,

> the client would cover that cookie with a signature.

Author: To be clear, I didn't look too closely at the federation text there.  The potential that things might be linked without going through the accepted processes (FedCM primarily, though also Storage Access on its own) concerned me enough to discount this.

Again, this is a case where the signing of the cookie allows for self-service.  Under any regime where you can obtain unpartitioned cookie access, this is pretty trivial.  The embedded SP (or IdP) gets the signatures it would have normally.  It can then tell the top-level site that everything is good.  At which point the top-level site can either set an ordinary cookie or it can do its own binding to a key in the TPM, remembering how it was blessed.

BTW, that whole idea about guessing secrets is a fine one, but orthogonal to this.  This is about ensuring that cookies cannot be stolen.  It is also probably an idea that makes cookie theft even more consequential, as it creates a way to stealthily track if you can guess correctly.

The use of the `Path` cookie parameter would ensure that
`Signed` cookies are only sent when requests are made to specific resources.
That would allow servers to limit how often they ask clients to generate signatures
by limiting how often the client requests the identified resources.
Those requests could be initiated through a redirect, fetch, or any other request, as needed.

The choice of what fields to include under the signature is very important.
It is not likely to be sufficient to just cover `Cookie` in the list of signed content in `Signature-Input`.
Including a date (the `created` parameter), the method (`@method`), and the URL (`@target-uri`)

> seem to be the minimum set of things that will prevent the signature from being reused.

Author: That doesn't argue either way for something like that.  Either something needs to be signed to achieve some goal, or not.  How something is spelled has no effect on that.

The only reasons I can see that would justify a distinct header are:

* to ensure that you don't get a signed cookie back from a client that doesn't understand the new feature
* to avoid all that `__Secure-`/Secure, HttpOnly, etc baggage that cookies have accrued over time

The first is pretty trivial to handle gracefully.  The second has a long history of failure.

It's possible that a more thorough security analysis will identifier other fields that need to be covered.

> ## Some Challenges

Reviewer: This section should be about how to accomplish various use cases, and I think it should start with a demo of how to satisfy the single threat model that DBSC currently covers: you trust the client's clock but not their timestamps. Then we can follow up with the extra flexibility that re-using cookies allows.

```suggestion
## Implementing various use cases

### Interactive proof if the client knows when to refresh
```

which might be able to recognize that the cookie is due for a refresh.
The response could forcibly expire the cookie
and force a fetch for a resource at `/foo/bar` in the background.
That fetch would cause the session refresh to occur,
without necessarily delaying the page load.

A similar approach is possibly under the alternative design.
Any resource could accept a cookie that the server wants to refresh
if that resource is less critical to protect.
This is more flexible because it is not tied to specific path prefixes.
It could instead be for other reasons,
such as whether the request is for protected information or actions.
Resources can then trigger asynchronous fetches to refresh cookies,
ahead of when any critical resources need to be fetched.

## Communicating Keys

Enrollment can almost be a side effect of creating and first use of a `Signed` cookie.
The only requirement here is that the browser learns what types of keys are acceptable to the server
and that the server learns the public key that the client uses.

Any `Set-Cookie` header that establishes a `Signed` cookie could list the key types in the `Signed` attribute,

> but the `Accept-Signature` field exists for negotiating the use of signature keys.

Reviewer: ```suggestion
but the [`Accept-Signature` field exists for negotiating the use of signature keys](https://www.rfc-editor.org/rfc/rfc9421.html#name-requesting-signatures).
```

without necessarily delaying the page load.

A similar approach is possibly under the alternative design.
Any resource could accept a cookie that the server wants to refresh
if that resource is less critical to protect.
This is more flexible because it is not tied to specific path prefixes.
It could instead be for other reasons,
such as whether the request is for protected information or actions.
Resources can then trigger asynchronous fetches to refresh cookies,
ahead of when any critical resources need to be fetched.

## Communicating Keys

Enrollment can almost be a side effect of creating and first use of a `Signed` cookie.
The only requirement here is that the browser learns what types of keys are acceptable to the server
and that the server learns the public key that the client uses.

Any `Set-Cookie` header that establishes a `Signed` cookie could list the key types in the `Signed` attribute,
but the `Accept-Signature` field exists for negotiating the use of signature keys.
The server should therefore use `Accept-Signature`.

The `Cookie` header that the browser subsequently sends will be signed.
That same message can include the public key from the key pair.

> That’s usually not something that can be included in the signature as defined in the curren.
> RFC.

Author: ```suggestion
That’s usually not something that can be included in the signature as defined in the current RFC.
```

  that it can use for protocol interactions,
* a way of identifying which resources use cookies that are provided using the protocol, plus
* the names of the cookies that can be produced.

The site is given the public key from the site-specific key pair that the browser holds.

In the proposed design,
the browser understands that when it makes a request to one of the resources that participates in the protocol,
it is expected to hold refreshed versions of the identified cookies.

These cookies are expected to have very short validity periods.
The browser is able to refresh those cookies automatically by interacting with the session resource.
The main part of the protocol is the interactions between the browser and that session resource.

Interactions with the session resource are a two-step process.
The first is a simple request that retrieves a fresh challenge.
The second posts a signature from the secret key over that challenge,
thereby proving to the server that the browser still has access to the key pair.
The response to the second request refreshes any of the affected cookies.

This adds two round trips of latency every time that a cookie refresh is needed.

> While some amount of delay is likely unavoidable, having two additional requests is fairly heavyweight.

Author: Yeah, I meant to cut this.

```suggestion
```

We have an alternative below that doesn't require an interactive exchange.
However, given that TPMs generally don't have a clock,
you can't use the clock to ensure freshness.
A non-interactive exchange might have been pre-generated by an attacker
who temporarily had access to the TPM, unless it contains fresh entropy from the server.
That's something we address in more detail in the alternative design below,
noting that the alternative offers servers more options to combine requests to reduce latency,

> where the proposal cannot.

Author: This is an important point in the requirement space to acknowledge.  That is, part of the point of the feature (perhaps the entire point) is to enable this sort of liveness test.  As the explainer isn't very clear about this aspect of the design, I thought it worth repeating.

> The proposal includes a redundant new session identifier field in requests.

Author: Fair.  My intent was to say that this was redundant with cookies.  The design isn't removing cookies, which could very easily identify a session, even when the session still requires a signature.

```suggestion
The proposal includes a new session identifier field in requests.
```

For that, we might define a new `Signature-Public-Key` field to carry the necessary information.
That would be a new piece of specification, but a simple one.
Something like the following would probably suffice:

```http
Signature-Public-Key: keyid="whatever", alg=ed25519, pubkey=:mc7Fpqi1aY/6...:
```

A possible alternative is to define a new `Signature` field parameter,
but that could be confused with `keyid`.

One nice thing about this is that the server can choose the `keyid` value
that is used when it sends `Accept-Signature`,
which provides the server with certainty about how keys are identified.
That key identifier is a form of cookie also;
or an extension to the information stored for the `Signed` cookie.
That means it needs to be cleared along with the cookie if someone asks the browser to clear state.
Of course, state clearing already requires that the key pair also be cleared.

The only thing remaining is to maybe avoid sending the public key
when the browser sends a `Signed` cookie subsequent to enrollment.
This can be as simple as remembering the last request that was made with that cookie.

> If the cookie has changed, or the last request made with that cookie received a 4x.

Author: ```suggestion
If the cookie has changed, or the last request made with that cookie received a 4xx
```
